File: apps/api/app/api/v1/endpoints/preview.py
from fastapi import APIRouter, HTTPException

from app.schemas.preview import (
    PreviewRequest,
    PreviewResponse,
    PreviewSamplesResponse,
    SendSamplesRequest,
    SendSamplesResponse,
)
from app.services.preview_samples_delivery import deliver_preview_samples
from app.services.preview_service import PreviewService

router = APIRouter()


@router.post("/search", response_model=PreviewResponse)
async def preview_search(payload: PreviewRequest) -> PreviewResponse:
    """Run a one-time mention search on br-analytics.ru for the landing preview."""
    service = PreviewService()
    try:
        return await service.search(payload.query)
    except Exception as exc:  # noqa: BLE001 — mapped to HTTP error at API boundary
        raise HTTPException(status_code=502, detail=f"br-analytics preview failed: {exc}") from exc


@router.post("/send-samples", response_model=SendSamplesResponse)
async def send_preview_samples(payload: SendSamplesRequest) -> SendSamplesResponse:
    """Email up to 3 sample mentions to the visitor (not shown on the public preview page)."""
    return await deliver_preview_samples(payload.sample_token, str(payload.email))


@router.get("/samples/{token}", response_model=PreviewSamplesResponse)
async def get_preview_samples(token: str) -> PreviewSamplesResponse:
    """Magic-link API: full sample texts for a limited-time token."""
    bundle = PreviewService().get_samples_by_token(token)
    if bundle is None:
        raise HTTPException(status_code=404, detail="Примеры не найдены или срок ссылки истёк.")
    return PreviewSamplesResponse(query=bundle.query, items=bundle.items)

File: apps/web/src/lib/api/preview.ts
import type {
  PreviewResponse,
  PreviewSamplesResponse,
  SendSamplesResponse,
} from "@/types/preview";

export async function previewSearch(query: string): Promise<PreviewResponse> {
  const response = await fetch("/api/v1/preview/search", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query }),
  });

  if (!response.ok) {
    throw new Error("Preview request failed");
  }

  return response.json() as Promise<PreviewResponse>;
}

export async function sendPreviewSamples(
  sampleToken: string,
  email: string,
): Promise<SendSamplesResponse> {
  const response = await fetch("/api/v1/preview/send-samples", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ sample_token: sampleToken, email }),
  });

  if (!response.ok) {
    throw new Error("Send samples failed");
  }

  return response.json() as Promise<SendSamplesResponse>;
}

export async function fetchPreviewSamples(token: string): Promise<PreviewSamplesResponse> {
  const response = await fetch(`/api/v1/preview/samples/${encodeURIComponent(token)}`);

  if (!response.ok) {
    throw new Error("Samples not found");
  }

  return response.json() as Promise<PreviewSamplesResponse>;
}
